akka-docs-dev/rst/scala/http/routing-dsl/directives/index.rst

```diff
@@ -0,0 +1,224 @@
+also use the ``|`` operator on directives. Here is yet another way to write the example:
+
+.. includecode2:: ../../../code/docs/http/scaladsl/server/DirectiveExamplesSpec.scala
+   :snippet: example-3
+
+Or better (without dropping down to writing an explicit ``Route`` function manually):
+
+.. includecode2:: ../../../code/docs/http/scaladsl/server/DirectiveExamplesSpec.scala
+   :snippet: example-4
+
+If you have a larger route structure where the ``(get | put)`` snippet appears several times you could also factor it
+out like this:
+
+.. includecode2:: ../../../code/docs/http/scaladsl/server/DirectiveExamplesSpec.scala
+   :snippet: example-5
+
+Note that, because ``getOrPut`` doesn't take any parameters, it can be a ``val`` here.
+
+As an alternative to nesting you can also use the `&` operator:
+
+.. includecode2:: ../../../code/docs/http/scaladsl/server/DirectiveExamplesSpec.scala
+   :snippet: example-6
+
+Here you can see that, when directives producing extractions are combined with ``&``, the resulting "super-directive"
+simply extracts the concatenation of its sub-extractions.
+
+And once again, you can factor things out if you want, thereby pushing the "factoring out" of directive configurations
+to its extreme:
+
+.. includecode2:: ../../../code/docs/http/scaladsl/server/DirectiveExamplesSpec.scala
+   :snippet: example-7
+
+This type of combining directives with the ``|`` and ``&`` operators as well as "saving" more complex directive
+configurations as a ``val`` works across the board, with all directives taking inner routes.
+
+Note that going this far with "compressing" several directives into a single one probably doesn't result in the most
+readable and therefore maintainable routing code. It might even be that the very first of this series of examples
+is in fact the most readable one.
+
+Still, the purpose of the exercise presented here is to show you how flexible directives can be and how you can
+use their power to define your web service behavior at the level of abstraction that is right for **your** application.
+
+
+Type Safety of Directives
+-------------------------
+
+When you combine directives with the ``|`` and ``&`` operators the routing DSL makes sure that all extractions work as
+expected and logical constraints are enforced at compile-time.
+
+For example you cannot ``|`` a directive producing an extraction with one that doesn't::
+
+    val route = path("order" / IntNumber) | get // doesn't compile
+
+Also the number of extractions and their types have to match up::
+
+    val route = path("order" / IntNumber) | path("order" / DoubleNumber)   // doesn't compile
+    val route = path("order" / IntNumber) | parameter('order.as[Int])      // ok
+
+When you combine directives producing extractions with the ``&`` operator all extractions will be properly gathered up::
+
+    val order = path("order" / IntNumber) & parameters('oem, 'expired ?)
+    val route =
+      order { (orderId, oem, expired) =>
+        ...
```

Comment:
maybe a comment here for newcomers to immediately see the types explained? 
`// (Int, String, Option[String])`?